execstats: ignore ComponentStats from other flows

asubiotto opened the pull request:
The TraceAnalyzer is initialized with a physical plan definition. Previously,
if it encountered unexpected stats in a trace, an error was returned. This
commit changes that behavior to only return an error if the unexpected stats
belong to a component of a flow the TraceAnalyzer was interested in. This is
done by matching the FlowID in the ComponentStats proto.

Release note: None

Fixes #60609

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60932)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60932#-:-MU9bavi7cZLgYPAO35X:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @RaduBerinde)

___
*[pkg/sql/execstats/traceanalyzer.go, line 81 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60932#-MU9a7RYCi4iZetRTp-N:-MU9a7RYCi4iZetRTp-O:b-9r9091) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc6be739a7f9e1203607a06fbfb6b6a424955b83/pkg/sql/execstats/traceanalyzer.go#L81)):*
> ```Go
> 		} else if util.CrdbTestBuild && !a.flowID.Equal(flow.FlowID.UUID) {
> 			panic(
> 				fmt.Sprintf(
> ```

nit: maybe `errors.AssertionFailedf`?
___
*[pkg/sql/execstats/traceanalyzer.go, line 165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60932#-MU9bKRt7pcSFpvFDPV0:-MU9bKRt7pcSFpvFDPV1:b5fbjrf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc6be739a7f9e1203607a06fbfb6b6a424955b83/pkg/sql/execstats/traceanalyzer.go#L165)):*
> ```Go
> 	// Annotate the maps with stats extracted from the trace.
> 	for component, componentStats := range m {
> 		if !component.FlowID.Equal(a.flowID.UUID) {
> ```

Not sure if it is important, but `uuid.UUID.Equal` will return `true` only if both arguments are exactly the same object (`uuid_wrapper.go:42`). 

If we want "deep" equality check, that we need to compare them as `[]byte`. (I remember looking at this last week.)


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60932)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde and @yuzefovich)

___
*[pkg/sql/execstats/traceanalyzer.go, line 165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60932#-MU9bKRt7pcSFpvFDPV0:-MU9epbW6Fsc_x1LJI8I:b-32rcfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc6be739a7f9e1203607a06fbfb6b6a424955b83/pkg/sql/execstats/traceanalyzer.go#L165)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Not sure if it is important, but `uuid.UUID.Equal` will return `true` only if both arguments are exactly the same object (`uuid_wrapper.go:42`). 

If we want "deep" equality check, that we need to compare them as `[]byte`. (I remember looking at this last week.)
</blockquote></details>

Done. thanks for the heads up


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60932)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)

___
*[pkg/sql/execstats/traceanalyzer.go, line 165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60932#-MU9bKRt7pcSFpvFDPV0:-MUA5ZPbCnr7z6ng-wrr:b-sutnfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc6be739a7f9e1203607a06fbfb6b6a424955b83/pkg/sql/execstats/traceanalyzer.go#L165)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Done. thanks for the heads up
</blockquote></details>

Are you sure? It looks like it compares two fixed-size arrays, which should be by value:
```
type UUID [Size]byte

func (u UUID) Equal(t UUID) bool {
	return u == t
}
```


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60932)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde and @yuzefovich)

___
*[pkg/sql/execstats/traceanalyzer.go, line 165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60932#-MU9bKRt7pcSFpvFDPV0:-MUA6jUt6HBbVhnyPG1p:b3iziv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc6be739a7f9e1203607a06fbfb6b6a424955b83/pkg/sql/execstats/traceanalyzer.go#L165)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Are you sure? It looks like it compares two fixed-size arrays, which should be by value:
```
type UUID [Size]byte

func (u UUID) Equal(t UUID) bool {
	return u == t
}
```
</blockquote></details>

Hm, indeed, my bad. I misunderstood the comment  - I read this part
```
// This method exists only to conform to the API expected by gogoproto's
// generated Equal implementations.
```
as "it doesn't work and exists only for conformance", and not as "it is redundant because == could be used instead".


<!-- Sent from Reviewable.io -->

asubiotto commented:
Updated

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60932#-:-MUAclVvAWfDTHIucj0i:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @yuzefovich)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60932)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/execinfrapb/api.go, line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60932#-MUAcuP66hIE13fwMzCT:-MUAcuP66hIE13fwMzCU:b-iu4iej) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b4283f35300f14aa7ddc665befdcd0f73fc6f14/pkg/sql/execinfrapb/api.go#L40)):*
> ```Go
> 
> // NilFlowID is the nil FlowID. Used to check for unset FlowIDs.
> var NilFlowID = FlowID{}
> ```

[nit] consider an `IsUnset()` method instead


<!-- Sent from Reviewable.io -->

asubiotto commented:
Interesting failure in `explain_analyze_plans` cc @RaduBerinde seems like we previously did not expect a `rows read from KV: 2 (16 B)` line in the query level stats for `EXPLAIN ANALYZE (DISTSQL) INSERT INTO child VALUES (1, (SELECT min(p) FROM parent))` and it now shows up. Going to `--rewrite` the test since it seems good but wanted to bring it up.

Commit:
execstats: ignore ComponentStats from other flows

The TraceAnalyzer is initialized with a physical plan definition. Previously,
if it encountered unexpected stats in a trace, an error was returned. This
commit changes that behavior to only return an error if the unexpected stats
belong to a component of a flow the TraceAnalyzer was interested in. This is
done by matching the FlowID in the ComponentStats proto.

Release note: None

RaduBerinde reviewed (commented):
Yeah, that's good. Components from the subquery were likely interfering.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60932#-:-MUAzOmA1RhgbTXiKRQy:b-aht13n)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @asubiotto and @yuzefovich)



<!-- Sent from Reviewable.io -->
